patches/tModLoader/Terraria/ModLoader/UI/ModBrowser/UIModBrowser.cs

```diff
@@ -184,103 +236,139 @@ public void BackClick(UIMouseEvent evt, UIElement listeningElement)
+			Interface.updateMessage.SetURL(_updateUrl);
+			Interface.updateMessage.SetAutoUpdateURL(_autoUpdateUrl);
+			Main.menuMode = Interface.updateMessageID;
+		}
+
+		lock (modSlugsToUpdateInstallInfo) {
+			foreach (var item in ModList.ReceivedItems.Where(
+				d => modSlugsToUpdateInstallInfo.Contains(d.ModDownload.ModName)
+			)) {
+				item.ModDownload.UpdateInstallState();
+				item.UpdateInstallInfo();
+			}
+			// @TODO: Shouldn't only delete processed slugs?
+			modSlugsToUpdateInstallInfo.Clear();
+		}
+
+		if (
+			(DebounceTimer is not null) &&
+			(DebounceTimer.ElapsedMilliseconds >= DEBOUNCE_MS)
+		) {
+			// No need to count more
+			DebounceTimer.Stop();
+			DebounceTimer = null;
+		}
 
-		foreach (var item in WorkshopHelper.QueryHelper.Items) {
-			_items.Add(new UIModDownloadItem(item));
+		if (UpdateNeeded) {
+			// Debounce logic
+			if (DebounceTimer is null) {
+				UpdateNeeded = false;
+				PopulateModBrowser();
+				DebounceTimer = new();
+				DebounceTimer.Start();
+			}
 		}
+	}
+
+	internal void PopulateModBrowser()
+	{
+		_firstLoad = false;
+
+		// Only called if using mod browser and not for modpacks
+		SpecialModPackFilter = null;
+		SpecialModPackFilterTitle = null;
+
+		// Since we could have used modpacks before fix the title if wrong
+		SetHeading(Language.GetText("tModLoader.MenuModBrowser"));
+
+		// Old data will be removed and old provider aborted when setting the new provider `ModList.SetProvider`
 
-		return UpdateNeeded = true;
+		// Asynchronous load the Mod Browser
+		ModList.SetEnumerable(SocialBackend.QueryBrowser(FilterParameters));
 	}
 
 	/// <summary>
 	///     Enqueues a list of mods, if found on the browser (also used for ModPacks)
 	/// </summary>
-	internal void DownloadMods(IEnumerable<string> modNames)
+	internal void DownloadMods(List<ModPubId_t> modIds)
 	{
-		var downloads = new List<ModDownloadItem>();
+		var downloadsQueried = SocialBackend.DirectQueryItems(new QueryParameters() { searchModIds = modIds.ToArray() });
 
-		foreach (string desiredMod in modNames) {
-			var mod = WorkshopHelper.QueryHelper.Items.FirstOrDefault(x => x.ModName == desiredMod);
-
-			if (mod == null) { // Not found on the browser
-				_missingMods.Add(desiredMod);
-			}
-			else if (mod.Installed == null || mod.HasUpdate) { // Found, add to downloads
-				downloads.Add(mod);
-			}
+		for (int i = 0; i < modIds.Count(); i++) {
+			if (downloadsQueried[i] == null)
+				//TODO: Would be nice if this was name/slug, not ID
+				_missingMods.Add(modIds[i].m_ModPubId);
 		}
 
+		var downloadShortList = ModDownloadItem.FilterOutInstalled(downloadsQueried);
+
 		// If no download detected for some reason (e.g. empty modpack filter), prevent switching UI
-		if (downloads.Count <= 0)
+		if (downloadShortList.Count() <= 0)
 			return;
 
-		WorkshopHelper.SetupDownload(downloads, Interface.modBrowserID);
+		SocialBackend.SetupDownload(downloadShortList.ToList(), Interface.modBrowserID);
```

Comment:
You never wait on this task, so if there are missing mods, it's going to replace the download UI with the info message UI